tempesta_fw/http_msg.c

```diff
@@ -935,34 +899,48 @@ int
 tfw_http_msg_add_data(TfwMsgIter *it, TfwHttpMsg *hm, TfwStr *field,
 		      const TfwStr *data)
 {
-	char *p;
-	unsigned int n_copy;
-
-	WARN_ON_ONCE(TFW_STR_DUP(data));
-	WARN_ON_ONCE(!TFW_STR_PLAIN(data));
+	unsigned int c_off = 0, c_size;
 
-	n_copy = min(data->len, (unsigned long)skb_tailroom(it->skb));
-	if (!n_copy)
-		return 0;
-
-	if (it->frag >= 0)
-		goto add_frag;
-
-	p = skb_put(it->skb, n_copy);
-	memcpy_fast(p, data->data, n_copy);
+	if (WARN_ON_ONCE(TFW_STR_DUP(data) || !TFW_STR_PLAIN(data)))
+		return -EINVAL;
+	if (WARN_ON_ONCE(it->frag >= skb_shinfo(it->skb)->nr_frags))
+		return -E2BIG;
 
-	if (__tfw_http_msg_add_str_data(hm, field, p, n_copy, it->skb))
-		return -ENOMEM;
+	while ((c_size = data->len - c_off)) {
+		char *p;
+		unsigned int n_copy, f_room;
+
+		if (it->frag >= 0) {
+			unsigned int f_size;
+			skb_frag_t *frag = &skb_shinfo(it->skb)->frags[it->frag];
+
+			f_size = skb_frag_size(frag);
+			f_room = PAGE_SIZE - frag->page_offset - f_size;
+			p = (char *)skb_frag_address(frag) + f_size;
+			n_copy = min(c_size, f_room);
+			skb_frag_size_add(frag, n_copy);
+			ss_skb_adjust_data_len(it->skb, n_copy);
+		} else {
+			f_room = skb_tailroom(it->skb);
+			n_copy = min(c_size, f_room);
+			p = skb_put(it->skb, n_copy);
+		}
```

Comment:
This is a copy&paste from `tfw_msg_write()` - can we reuse the code, e.g. by moving some of the code to a helper function?